### ISSUE TITLE
Topic Models for scVI - ProdLDA

### ISSUE THREAD

**Issue by Edouard360** [YosefLab/scVI-dev#38]
The LDA model can be decomposed into two parts: the Dirichlet for the mixture of topics, and the multinomials for the words distribution over each topic, which can be viewed as a matrix BETA of parameters to learn. The [prodLDA](https://arxiv.org/pdf/1703.01488.pdf) model removes the usual hypothesis that each word must be issued from a single topic - ie. that we draw a multinomial over the latent parameter $\theta \sim Dir(\alpha)$. 

Instead, they make the simplifying assumption that each word itself is drawn from a mixture of topics, ie. $w_n \sim Mult(\beta \theta)$ (dot product).

This might (?) be plugged-in in our single-cell case, if we consider topics to be cell-types, and then $\beta$ is a matrix where each column represents a cell type, and each line entry corresponds to the normalized expression of the gene: what is referred to $w_{n,g}$ in Romain's paper.

So instead of having $p(w_{n,g}|z)$ modeled by an arbitrary neural network, we enforce the normalized expression $w_{n,g}$ to be a mixture of normalized expression for cell type 1 $\beta_{1..G,1}$, cell type 2 $\beta_{1..G,2}$, until cell-type $K$ with $K=n_{\text{latent}}=10$ here ( cell type K ).

Taking the scvi code, I think it would amount to do, in decoder initialization:

`        self.BETA = nn.Parameter(torch.Tensor(self.n_input, self.n_latent)) # Matrix Beta equivalent
`

And then in decoder forward pass:

` theta = F.softmax(z, dim=-1)`

`px_scale = F.linear(theta, nn.Softmax(dim=0)(self.BETA), bias=None)`

Then we can leave the Gaussian prior to N(0,I), which corresponds to a Dirichlet prior of $\alpha = K/K-1$. After 250 epochs, Log-likelihood Train is 1650 and Log-likelihood Test is 1490, much worse than the scVI model.

**Comment by Edouard360** [comment]
# Experiment

So here's the [em/prodlda](/YosefLab/scVI-dev/tree/em/prodlda) I have been working on.

The context of the experiment is the following:

<img width="817" alt="adaptative lda" src="https://user-images.githubusercontent.com/15527397/38959203-975280f4-4314-11e8-9673-ac25fd0a64c3.png">

In the code the Beta is a parameter in the model to learn. It is a matrix of size (n_words * n_topics) or, in our case (n_genes * n_cell_types), where each entry corresponds to the normalized gene expression for that particular cell-types. 

https://github.com/YosefLab/scVI-dev/blob/1eccf82fa9a69aea1cdbb8fe4ae1f7142eb52ed9/scvi/scvi.py#L199-L202

Therefore each column represents the normalized expression for all genes of a specific cell-type. 

In the latent space, z no longer represents the latent variable as before, but it is rather the logit for the logistic normal approximation of the Dirichlet. To this effect, we apply a softmax on the dimensions of z. 

https://github.com/YosefLab/scVI-dev/blob/1eccf82fa9a69aea1cdbb8fe4ae1f7142eb52ed9/scvi/scvi.py#L229-L232

Now the normalized genes expression of each cell is a convex combination of the representative vectors of normalized genes expression for each cell-type, as implemented with `F.Linear(, bias=False)`.

This is to compare with the original way we get this normalized expression from the latent variable z, and linear layers at the end of which we apply a softmax.

https://github.com/YosefLab/scVI-dev/blob/1eccf82fa9a69aea1cdbb8fe4ae1f7142eb52ed9/scvi/scvi.py#L221

https://github.com/YosefLab/scVI-dev/blob/1eccf82fa9a69aea1cdbb8fe4ae1f7142eb52ed9/scvi/scvi.py#L236-L237

**NB:**

Actually, **topics might not necessarily correspond to cell-types**, but for instance could be interpreted as a cell-functionalities; and biologically, such a model would represent a cell as a "mixture of functionalities".

# Results

The final Log-likelihood is around 1500 and the clusters repartition is as follows for respectively, 3, 5, and 10 latent spaces:

## n_clusters = 3 (= n_cell_types / n_cell_functionalities)

0 epochs (before training)
![clustering_3_0](https://user-images.githubusercontent.com/15527397/38959875-2b3b7eea-4317-11e8-8747-480ee7a804cd.png)

250 epochs (after training)
![clustering_3_240](https://user-images.githubusercontent.com/15527397/38959876-2e2689f6-4317-11e8-98b2-d1aef3df7056.png)

## n_clusters = 5 (= n_cell_types / n_cell_functionalities)

0 epochs (before training)
![clustering_5_0](https://user-images.githubusercontent.com/15527397/38959929-598f66c6-4317-11e8-95a7-26d0613a1039.png)

250 epochs (after training)
![clustering_5_180](https://user-images.githubusercontent.com/15527397/38959936-621fe522-4317-11e8-8aa4-fabcb695224c.png)

## n_clusters = 10 (= n_cell_types / n_cell_functionalities)

0 epochs (before training)
![clustering_10_0](https://user-images.githubusercontent.com/15527397/38959950-6f1d8694-4317-11e8-98a0-441dfda5041c.png)

250 epochs (after training)
![clustering_10_150](https://user-images.githubusercontent.com/15527397/38959961-77c960ec-4317-11e8-9943-b0de2d3f5c29.png)


As we increase n_clusters, we realize that most topic are "garbage", and don't contain much information. Maybe some clever initialization of the BETA matrix with the mean gene level expression for each cell types would help converge towards better clustering results.

In the meantime, we will focus on semi-supervised and this closes #38.

**Comment by Edouard360** [comment]
Close #38

**Comment by jeff-regier** [comment]
Nice work Eddie. Thanks for the report.
